File: morphac/utils/test/integrator_utils_test.cc
#include "utils/include/integrator_utils.h"

#include "constructs/include/control_input.h"
#include "constructs/include/state.h"
#include "gtest/gtest.h"
#include "math/numeric/include/all_integrators.h"
#include "mechanics/models/include/diffdrive_model.h"

namespace {

using Eigen::VectorXd;
using morphac::constructs::ControlInput;
using morphac::constructs::State;
using morphac::math::numeric::EulerIntegrator;
using morphac::math::numeric::Integrator;
using morphac::math::numeric::IntegratorType;
using morphac::math::numeric::MidPointIntegrator;
using morphac::math::numeric::RK4Integrator;
using morphac::mechanics::models::DiffdriveModel;
using morphac::utils::IntegratorFromType;

TEST(IntegratorUtilsTest, Construction) {
  // Test the construction for different types. Make sure the right integrator
  // is constructed for the given type.
  // Also check to make sure that the integrator is usable.
  DiffdriveModel model{1., 1.};
  auto integrator = IntegratorFromType(IntegratorType::kEulerIntegrator, model);
  ASSERT_TRUE(dynamic_cast<EulerIntegrator*>(integrator.get()) != nullptr);
  State integrated_state = integrator->Step(State(3, 0), ControlInput(2), 0.05);
  ASSERT_TRUE(integrated_state.get_data().isApprox(VectorXd::Zero(3)));

  integrator = IntegratorFromType(IntegratorType::kMidPointIntegrator, model);
  ASSERT_TRUE(dynamic_cast<MidPointIntegrator*>(integrator.get()) != nullptr);
  integrated_state = integrator->Step(State(3, 0), ControlInput(2), 0.05);
  ASSERT_TRUE(integrated_state.get_data().isApprox(VectorXd::Zero(3)));

  integrator = IntegratorFromType(IntegratorType::kRK4Integrator, model);
  ASSERT_TRUE(dynamic_cast<RK4Integrator*>(integrator.get()) != nullptr);
  integrated_state = integrator->Step(State(3, 0), ControlInput(2), 0.05);
  ASSERT_TRUE(integrated_state.get_data().isApprox(VectorXd::Zero(3)));
}

}  // namespace

int main(int argc, char** argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: morphac/utils/python/test/integrator_utils_test.py
import numpy as np
import pytest

from morphac.constructs import State, ControlInput
from morphac.math.numeric import (
    IntegratorType,
    EulerIntegrator,
    MidPointIntegrator,
    RK4Integrator,
)
from morphac.mechanics.models import DiffdriveModel
from morphac.utils.integrator_utils import integrator_from_type


@pytest.fixture
def generate_integrator_list():

    integrator_type_list = [
        IntegratorType.EULER_INTEGRATOR,
        IntegratorType.MID_POINT_INTEGRATOR,
        IntegratorType.RK4_INTEGRATOR,
    ]
    integrator_class_list = [EulerIntegrator, MidPointIntegrator, RK4Integrator]

    return integrator_type_list, integrator_class_list


def test_construction(generate_integrator_list):

    integrator_type_list, integrator_class_list = generate_integrator_list
    kinematic_model = DiffdriveModel(1.0, 1.0)

    # Making sure that the right integrator is constructed.
    for integrator_type, integrator_class in zip(
        integrator_type_list, integrator_class_list
    ):
        integrator = integrator_from_type(
            integrator_type=integrator_type, kinematic_model=kinematic_model
        )
        assert isinstance(integrator, integrator_class)

        # Also making sure that the integrator is usable.
        assert np.allclose(
            integrator.step(State(3, 0), ControlInput(2), 0.05).data, [0.0, 0.0, 0.0]
        )


def test_construction_with_temporary(generate_integrator_list):
    # Test the function with a temporarily created KinematicModel object.
    # This is to test for any cpp lifetime management weirdness.
    integrator_type_list, integrator_class_list = generate_integrator_list

    # Making sure that the right integrator is constructed.
    for integrator_type, integrator_class in zip(
        integrator_type_list, integrator_class_list
    ):
        integrator = integrator_from_type(integrator_type, DiffdriveModel(1.0, 1.0))

        assert isinstance(integrator, integrator_class)

        # Also making sure that the integrator is usable.
        assert np.allclose(
            integrator.step(State(3, 0), ControlInput(2), 0.05).data, [0.0, 0.0, 0.0]
        )
